src/main/java/store/domain/Receipt.java

```diff
@@ -0,0 +1,120 @@
+        int bundle = promotion.getBuy() + promotion.getGet();
+        int giveAwayQuantity = Math.min(quantity, product.getQuantity()) / bundle;
+        if (giveAwayQuantity != 0) {
+            giveAwayHistory.put(product, giveAwayQuantity);
+        }
+    }
+
+    public void applyMembershipDiscount(Promotions promotions) {
+        for (Map.Entry<Product, Integer> purchaseInfo : purchaseHistory.entrySet()) {
+            Product product = purchaseInfo.getKey();
+            int quantity = purchaseInfo.getValue();
+            if (!hasGiveAway(product.getName())) {
+                membershipApplicableAmount += product.getPrice() * quantity;
+                continue;
+            }
+            int promotionApplicableQuantity = calculatePromotionApplicableQuantity(product, promotions);
+            membershipApplicableAmount += product.getPrice() * (quantity - promotionApplicableQuantity);
+        }
+    }
+
+    private int calculatePromotionApplicableQuantity(Product product, Promotions promotions) {
+        int giveAwayQuantity = findGiveAwayQuantity(product.getName());
+        Promotion promotion = promotions.findPromotion(product.getPromotionName());
+        int bundle = promotion.calculateBundle();
+        return bundle * giveAwayQuantity;
+    }
+
+    private boolean hasGiveAway(String productName) {
+        for (Map.Entry<Product, Integer> giveAwayInfo : giveAwayHistory.entrySet()) {
+            Product product = giveAwayInfo.getKey();
+            if (productName.equals(product.getName())) {
+                return true;
+            }
+        }
+        return false;
+    }
+
+    private int findGiveAwayQuantity(String productName) {
+        for (Map.Entry<Product, Integer> giveAwayInfo : giveAwayHistory.entrySet()) {
+            Product product = giveAwayInfo.getKey();
+            if (productName.equals(product.getName())) {
+                return giveAwayInfo.getValue();
+            }
+        }
+        return 0;
+    }
+
+    public int calculateTotalPurchaseAmount() {
+        int totalPurchaseAmount = 0;
+        for (Map.Entry<Product, Integer> giveAway : purchaseHistory.entrySet()) {
+            Product product = giveAway.getKey();
+            int quantity = giveAway.getValue();
+            totalPurchaseAmount += product.getPrice() * quantity;
+        }
+        return totalPurchaseAmount;
+    }
+
+    public int calculateTotalQuantity() {
+        int totalQuantity = 0;
+        for (Integer quantity : purchaseHistory.values()) {
+            totalQuantity += quantity;
+        }
+        return totalQuantity;
+    }
+
+    public int calculatePromotionDiscount() {
+        int promotionDiscount = 0;
+        for (Map.Entry<Product, Integer> giveAway : giveAwayHistory.entrySet()) {
+            Product product = giveAway.getKey();
+            int quantity = giveAway.getValue();
+            promotionDiscount += product.getPrice() * quantity;
+        }
```

Comment:
stream을 이용하면 더 깔끔해질 것 같아요!
```suggestion
        promotionDiscount = giveAwayHistory.entrySet().stream()
    .mapToInt(entry -> entry.getKey().getPrice() * entry.getValue())
    .sum();
```